src/tactic/protected.lean

```diff
@@ -0,0 +1,73 @@
+/-
+Copyright (c) 2020 Chris Hughes. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Hughes
+-/
+import tactic.suggest tactic.find
+/-!
+## `protected` and `protect_proj` user attributes
+
+`protected` is an attribute to protect a declaration
+If a declaration `foo.bar` is marked protected, then it must be referred to
+by its full name `foo.bar`, even when the `foo` namespace is open."
+
+`protect_proj` attribute to protect the projections of a structure.
+If a structure `foo` is marked with the `protect_proj` user attribute, then
+all of the projections become protected.
+
+`protect_proj without bar baz` will protect all projections except for bar and baz.
+-/
+
+namespace tactic
+/-- Attribute to protect a declaration
+    If a declaration `foo.bar` is marked protected, then it must be referred to
+    by its full name `foo.bar`, even when the `foo` namespace is open.. -/
+@[user_attribute] meta def protected_attr : user_attribute :=
+{ name := "protected",
+  descr := "Attribute to protect a declaration
+    If a declaration `foo.bar` is marked protected, then it must be referred to
+    by its full name `foo.bar`, even when the `foo` namespace is open.",
+  after_set := some (λ n _ _, mk_protected n)  }
+
+add_tactic_doc
+{ name        := "protected",
+  category    := doc_category.attr,
+  decl_names  := [`tactic.protected_attr],
+  tags        := ["user attribute"] }
+
+/-- Tactic that is executed when a structure is marked with the `protect_proj` attribute -/
+meta def protect_proj_tac (n : name) (l : list name) : tactic unit :=
+do env ← get_env,
+option.cases_on (env.structure_fields_full n)
+  (fail "protect_proj failed: declaration is not a structure")
+  (λ fields, fields.foldl
+    (λ t field, cond (l.foldl (λ b m, bor b (m.is_suffix_of field)) ff)
+      t
+      (t >> tactic.mk_protected field))
+    skip)
```

Comment:
I found the implementation here a little hard to read; I think this is equivalent but I didn't test. 

```suggestion
do env ← get_env,
match env.structure_fields_full n with
| none := fail "protect_proj failed: declaration is not a structure"
| some fields := fields.mmap' $ λ field,
    when (l.all $ λ m, bnot $ m.is_suffix_of field) $ mk_protected field
end
```